patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch

```diff
@@ -0,0 +1,695 @@
++            HRESULT hr;
++
++            // Checks if user granted permission to access the camera.
++            THR(async_status == AsyncStatus::Completed ? S_OK : E_ACCESSDENIED);
++
++          Cleanup:
++            ::SetEvent(
++                event_wait_for_media_capture_async_action_completion.Get());
++
++            return hr;
++          })
++          .Get()));
++
++  THR(::WaitForSingleObjectEx(
++          event_wait_for_media_capture_async_action_completion.Get(), INFINITE,
++          FALSE) == WAIT_OBJECT_0
++          ? S_OK
++          : E_FAIL);
++
++  // User granted access, good to go.
++  ReleaseInterface(ppMediaCapture);
++  *ppMediaCapture = media_capture.Detach();
++
++Cleanup:
++  return hr;
++}
++
+ uint32_t SafelyComputeMediaRatio(IMediaRatio* ratio_no_ref) {
+   HRESULT hr;
+   uint32_t numerator, denominator, media_ratio = 0;
+@@ -70,6 +226,31 @@ VideoType ToVideoType(const HString& sub_type) {
+   return converted_type;
+ }
+ 
++HRESULT GetMainViewDispatcher(ICoreDispatcher** main_view_dispatcher) {
++  HRESULT hr;
++
++  ComPtr<IActivationFactory> activation_factory;
++  ComPtr<ICoreApplication> coreApplication;
++  ComPtr<ICoreImmersiveApplication> immersiveApplication;
++  ComPtr<ICoreApplicationView> applicationView;
++  ComPtr<ICoreWindow> coreWindow;
++
++  THR(GetActivationFactory(
++      HStringReference(
++          RuntimeClass_Windows_ApplicationModel_Core_CoreApplication)
++          .Get(),
++      &activation_factory));
++
++  THR(activation_factory.As<ICoreApplication>(&coreApplication));
++  THR(coreApplication.As<ICoreImmersiveApplication>(&immersiveApplication));
++  THR(immersiveApplication->get_MainView(&applicationView));
++  THR(applicationView->get_CoreWindow(&coreWindow));
++  THR(coreWindow->get_Dispatcher(main_view_dispatcher));
++
++Cleanup:
++  return hr;
++}
++
+ HRESULT WaitForAsyncAction(ComPtr<IAsyncAction> async_action) {
+   HRESULT hr, hr_async_error;
+   const DWORD timeout_ms = 2000;
+diff --git a/modules/video_capture/windows/help_functions_winrt.h b/modules/video_capture/windows/help_functions_winrt.h
+index 0216d9b0ce..6b81bfc977 100644
+--- a/modules/video_capture/windows/help_functions_winrt.h
++++ b/modules/video_capture/windows/help_functions_winrt.h
+@@ -13,6 +13,7 @@
+ 
+ #include <windows.foundation.h>
+ #include <windows.media.mediaproperties.h>
++#include <windows.ui.core.h>
+ #include <wrl/client.h>
+ #include <wrl/event.h>
+ #include <wrl/wrappers/corewrappers.h>
+@@ -20,15 +21,51 @@
+ #include <cassert>
+ 
+ #include "modules/video_capture/video_capture_defines.h"
++#include "rtc_base/logging.h"
+ 
+ // Evaluates an expression returning a HRESULT.
+ // It jumps to a label named Cleanup on failure.
+-#define THR(expr)   \
+-  do {              \
+-    hr = (expr);    \
+-    if (FAILED(hr)) \
+-      goto Cleanup; \
++#ifdef _DEBUG
++#define THR(expr)                                           \
```

Comment:
IIRC the problem with `#define MACRO (...) if (FAILED(...)) return;` was that you couldn't break on the `return` i.e. on the failure. This has a `__debugbreak()` and it seems to be used mostly an assert rather than to validate input/handle expected error conditions, so maybe it's fine?

Reply:
> The alternative would make the code harder to read.

The page you linked only contains examples which do _not_ use such a macro in any of the 4 error handling patterns.

> A considerable part of the OS components (which one could call as a large project...) and several third-party COM components rely on macros like this one.

This code is old, I do not consider it a good reference on modern coding practices. Again, I have some experience debugging through the code you mention, especially via crash dumps, and being unable to set breakpoints makes things unnecessarily difficult. As Filippo pointed out the `__debugbreak()` here alleviates the issue though, so let's say this is OK.